### ISSUE TITLE
Kubernetes: Planned Changes / Deprecation / Replacement

### ISSUE THREAD

**Issue by bradrydzewski** [drone/drone-runtime#69]
__IMPORTANT__ We are no longer developing, improving or supporting the experimental Kubernetes runtime. You should instead use the traditional Docker runtime and agents. Read below for more details.

---
I consider the first iteration of the experimental kubernetes runtime to have been successful from a learning perspective, but a failure in terms of design and implementation. One thing that has become clear is that the current .drone.yml file was designed for Docker and we have had difficulty mapping these concepts to Kubernetes. It works but I do not believe we should build the future of Drone with this implementation as the foundation, as I mentioned in https://github.com/drone/drone-runtime/issues/53#issuecomment-482473370. Second is that many users have raised concerns about creating per-pipeline namespaces.

I am proposing that we create Custom Pipeline Definitions (similar to Kubernetes Customer Resource Definitions) that allow anyone to create a custom Pipeline workflow and custom pipeline syntax.

This is roughly outlined here:
https://github.com/drone/drone/issues/2680

The kubernetes runtime and yaml syntax will be moved to a separate repository where it can evolve independently, at its own pace. The yaml syntax will be slightly re-worked and optimized for kubernetes. I anticipate the following changes:

The type parameter will be required:

```diff
kind: pipeline
+type: kubernetes
```

Service containers should be launched in the same pod so that we can avoid creating Kubernetes service objects (one of the main reasons we used custom namespaces). Services will be available at localhost instead hostnames. By eliminating services we can eliminate the per-pipeline namespace requirement.

```diff
steps:
  - name: build
    image: golang
    commands:
    - go build
    - go test
+   services:
+   - name: redis
+     image: redis
```

Expanding on the above example, we may want to consider representing pipeline steps more like pods, where multiple containers including services can run in parallel. Not quite sure what the syntax would look like, but perhaps something like this:

```
steps:
- containers:
  - name: build
    image: golang
    commands:
    - go build

- containers:
  - name: test
    image: golang
    commands:
    - go test -v

  - name: redis
    image: redis
```

If we do something like above, we could define per-pod resource requirements.

```
steps:
- containers:
  - name: build
    image: golang
    commands:
    - go build
  resources:
    limit:
      memory: 500
      cpu: 1000

- containers:
  - name: test
    image: golang
    commands:
    - go test -v

  - name: redis
    image: redis
```

And we would probably move the `when` clause from being a container attribute to a pod attribute:

```
steps:
- containers:
  - name: build
    image: golang
    commands:
    - go build
  when:
    branch: [ master ]
```

In summary, I think the experimental Kubernetes runtime has helped us figure out what is working and what is not. We will plan on scrapping the experimental runtime in favor of a v2 implementation with the goal of a native Kubernetes runtime that is designed and optimized for Kubernetes.

**Comment by bradrydzewski** [comment]
example that shows all supported fields:

```
kind: pipeline
type: kubernetes
name: default

metadata:
  namespace: default
  labels:
    key1: value1
    key2: value2
  annotations:
    key1: value1
    key2: value2

steps:
- jobs:
  - name: build
    image: golang
    privileged: false
    pull: if-not-exists
    failure: ignore
    user: root
    commands:
    - go build
    - go test
  services:
  - name: redis
    image: redis
    environment:
      KEY1: value1
    entrypoint: [ /bin/redis ]
    command: [ --some-flag ]
    working_dir: /tmp
    devices:
    - name: data
      path: /dev/xda1
    mounts:
    - path: /test-ebs
      name: test-volume
    resources:
      requests:
        memory: "64Mi"
        cpu: "250m"
      limits:
        memory: "128Mi"
        cpu: "500m"
  when:
    branch:
    - master
    - develop

- jobs:
  - name: test
    image: golang
    commands:
    - go build
    - go test

volumes:
- name: test-volume
  aws_elastic_block_store:
      volume_id: <volume-id>
      fs_type: ext4

node_selector:
  key1: value1
  key2: value2

tolerations:
- key: "key"
  operator: "Equal"
  value: "value"
  effect: "NoSchedule"

image_pull_secrets:
- secret1
```

questions are

1. how to handle secrets? how to differentiate between drone secrets and kubernetes secrets?
2. how to ensure the pipeline is scheduled to run on a machine with enough resources

**Comment by stevecrozz** [comment]
Regarding question 2, in my ideal world I would actually not need to guarantee that a whole pipeline runs on a single machine. I'd rather let kube figure out where to schedule pods for me based on their own resource constraints and the resources available which is one area where kube really shines. This would allow me to efficiently schedule even very large pipelines across whatever spare resources are available.

**Comment by tboerger** [comment]
@stevecrozz this kind of scheduling requires also some other things, like readwritemany volumes. Maybe this should be a config option for the server as not all clusters provide these storage classes.

**Comment by schmitch** [comment]
@tboerger why does I need a readwritemany volume? currently considering a pipeline with multiple jobs/services
currently postgres or redis does not need to run on the same node as the rest of the job and also multiple jobs don't need the same storage.

so consider the following:
- job1
- job2
- redis
- postgres

all 4 resources could run on another maschine if done correctly. (btw. gitlab-ci won't do this)

in gitlab-ci considering:
- job1 + redis
- job2 + postgres

only job1+redis / job2+postgres can on different nodes since they are a pod. but thats only a limitation since you should call your services with localhost and not with a service. if a good service handling is created one could easily use that.

**Comment by bradrydzewski** [comment]
@tboerger perhaps we should just close this and let the community create kubernetes runners using our [custom runner definitions](https://github.com/drone/drone/issues/2680) and our runner framework?  Is there any reason to provide an official Drone kubernetes runner? So far we seem to lack consensus on how to design and build a kubernetes runner, which perhaps means we are not ready to have an official implementation.

**Comment by stevecrozz** [comment]
I'm not sure there's no reason to provide an official kube runner, but there are competing priorities and as much as I love kube, it would be a bummer to see precious core maintainer resources spent on maintaining a runner for every process scheduler under the sun. So I tend to agree.

**Comment by bradrydzewski** [comment]
closing this issue. old kubernetes runner scrapped. new kubernetes runner moved to [drone-runners/drone-runner-kube](https://github.com/drone-runners/drone-runner-kube).

**Comment by jesselang** [comment]
> closing this issue. old kubernetes runner scrapped. new kubernetes runner moved here:
> #69

Your comment points to this issue itself. Is that intentional? 😕

**Comment by bradrydzewski** [comment]
oops, wrong link.  the new kubernetes runner can be found here:
https://github.com/drone-runners/drone-runner-kube

documentation should be live some time tomorrow at docs.drone.io.  They syntax is very close to the docker runner syntax, with the exception that you have to specify `type: kubernetes`.  There are also some minor behavior differences. The most notable is that services are available at localhost instead of a custom hostname.

For example:

```text
kind: pipeline
type: kubernetes
name: default

clone:
  disable: true

steps:
- name: test
  pull: if-not-exists
  image: redis
  commands:
  - sleep 5
  - redis-cli -h localhost ping
  - redis-cli -h localhost set FOO bar
  - redis-cli -h localhost get FOO

services:
- name: redis
  pull: if-not-exists
  image: redis
```

**Comment by bradrydzewski** [comment]
link to first-draft of kubernetes runner documentation:
https://kube-runner.docs.drone.io/
